### ISSUE TITLE
VirtualPlatformState getParticipatingModules order is inconsistent between builds

### ISSUE THREAD

**Issue by pinkhorror** [gradle/gradle#11714]
### Expected Behavior

We expect the dependency graph builder to produce nodes in a consistent order, for the classpath to be in a consistent order, for compilation to count as up-to-date when nothing changes.

### Current Behavior

Rerunning a target such as "test" on a project with this problem could result in recompiling the main code, the test code, or both.

### Context

We're trying to skip the parts that do not need to be rebuilt so we can run the targets we're interested in faster.

### Steps to Reproduce 

I haven't fully figured out everything that goes into this. I think I key part is that we have a plugin that ends up calling ComponentMetadataDetails.belongsTo during the dependency graph builder, that's set up something like this (very simplified):
```
project.getConfigurations().configureEach(conf -> {
    conf.getIncoming().beforeResolve(c -> {
        if (!handlerIsAlreadyApplied(project)) {
            setHandlerAsApplied(project);
            project.getDependencies().components(handler -> {
                handler.all(details -> {
                    String platformName = getAlignPlatformFor(details.getId());
                    if (null != platformName) {
                        details.belongsTo("align-platform-name:" + platformName + ":" + details.getId().getVersion());
                    }
                });
            });
        }
    });
});
```

### Your Environment

The gradle we're using is based off 5.6.3. this also happened on our 5.3.x version. I've fixed the issue in-house by not using a hash set within VirtualPlatformState, and I've been asked to write an issue for your tracker, but all the gradle work I've ever done has been with this modified version and I'm not sure how to even make a test project for you with regular gradle, without the company-specific core modifications and plugins. I think there's a chance we've caused the issue ourselves, or that the plugin is setting up the handler above the wrong way.

**Comment by sonicsnes** [comment]
Appears that this is probably still an issue in master:

getCandidateVersions sorts the participatingModules before using them (making it consistent): https://github.com/gradle/gradle/blob/master/subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/builder/VirtualPlatformState.java#L91

but getParticipatingModules is still used with the raw unsorted HashSet to do things like create the `PotentialEdge`s, which results in an inconsistent resolve: https://github.com/gradle/gradle/blob/49292195fab013503512b86b0051191fcfdb9ad3/subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/builder/LenientPlatformDependencyMetadata.java#L182

**Comment by ljacomet** [comment]
This was fixed through #13164 which was caused by the same underlying issue.
